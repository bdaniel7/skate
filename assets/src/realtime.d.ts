import {
  BlockId,
  DirectionId,
  RouteId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule.d"

import { HeadwaySpacing } from "./models/vehicleStatus"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  scheduledTimepointStatus: VehicleTimepointStatus
}

export type SourceId = string

export type RunId = string

export interface Vehicle {
  id: VehicleId
  label: string
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId
  routeId: RouteId
  tripId: TripId
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string
  operatorName: string
  bearing: number
  blockId: BlockId
  headwaySecs: number | null
  headwaySpacing: HeadwaySpacing | null
  previousVehicleId: string
  scheduleAdherenceSecs: number
  scheduleAdherenceString: string
  scheduledHeadwaySecs: number
  isLayingOver: boolean
  layoverDepartureTime: number | null
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  isOnRoute: boolean
}

export type VehicleOrGhost = Vehicle | Ghost

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId
  stopName: string
}

export interface VehicleScheduledLocation {
  directionId: DirectionId
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}

export interface VehiclesForRoute {
  onRouteVehicles: Vehicle[]
  incomingVehicles: Vehicle[]
  ghosts: Ghost[]
}
