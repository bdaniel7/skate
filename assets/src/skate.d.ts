export interface Route {
  id: RouteId
}

export type RouteId = string
