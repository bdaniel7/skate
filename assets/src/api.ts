import { ByRouteId, Route, RouteId, TimepointId } from "./skate.d"

interface RoutesResponse {
  data: Route[]
}

interface TimepointsResponse {
  data: ByRouteId<TimepointId[]>
}

const checkResponseStatus = (response: Response) => {
  if (response.status !== 200) {
    throw new Error(`Response error: ${response.status}`)
  }
  return response
}

const parseJson = (response: Response) => response.json()

export const fetchRoutes = (): Promise<Route[]> =>
  fetch("/api/routes")
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: routes }: RoutesResponse) => routes)
    .catch(error => {
      // tslint:disable-next-line: no-console
      console.error(error)
      throw error
    })

export const fetchTimepointsForRoutes = (
  routeIds: RouteId[]
): Promise<ByRouteId<TimepointId[]>> => {
  const queryParams = routeIds
    .map((routeId: RouteId) => "route_id[]=" + routeId)
    .join("&")
  return fetch(`/api/timepoints?${queryParams}`)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(
      ({ data: timepointsByRouteId }: TimepointsResponse) => timepointsByRouteId
    )
    .catch(error => {
      // tslint:disable-next-line: no-console
      console.error(error)
      throw error
    })
}
