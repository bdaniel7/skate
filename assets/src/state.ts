import { Dispatch as ReactDispatch } from "react"
import { RouteId } from "./skate.d"

export interface State {
  selectedRouteIds: RouteId[]
}

export const initialState: State = {
  selectedRouteIds: [],
}

interface SelectRoutesAction {
  type: "SELECT_ROUTES"
  payload: {
    routeIds: RouteId[]
  }
}

export const selectRoute = (routeId: RouteId): SelectRoutesAction => ({
  type: "SELECT_ROUTES",
  payload: { routeIds: [routeId] },
})

export const selectRoutes = (routeIds: RouteId[]): SelectRoutesAction => ({
  type: "SELECT_ROUTES",
  payload: { routeIds },
})

interface DeselectRoutesAction {
  type: "DESELECT_ROUTES"
  payload: {
    routeIds: RouteId[]
  }
}

export const deselectRoute = (routeId: RouteId): DeselectRoutesAction => ({
  type: "DESELECT_ROUTES",
  payload: { routeIds: [routeId] },
})

export const deselectRoutes = (routeIds: RouteId[]): DeselectRoutesAction => ({
  type: "DESELECT_ROUTES",
  payload: { routeIds },
})

type Action = SelectRoutesAction | DeselectRoutesAction

export type Dispatch = ReactDispatch<Action>

export const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "SELECT_ROUTES":
      return {
        ...state,
        selectedRouteIds: state.selectedRouteIds.concat(
          action.payload.routeIds
        ),
      }
    case "DESELECT_ROUTES":
      return {
        ...state,
        selectedRouteIds: state.selectedRouteIds.filter(
          id => !action.payload.routeIds.includes(id)
        ),
      }
    default:
      return state
  }
}
