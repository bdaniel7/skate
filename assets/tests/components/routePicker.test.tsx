import { shallow } from "enzyme"
import React from "react"
import renderer from "react-test-renderer"
import RoutePicker from "../../src/components/routePicker"
import { Route, RouteId } from "../../src/skate"
import { deselectRoute, selectRoute } from "../../src/state"

const emptyDispatch = () => undefined

test("renders a list of routes", () => {
  const routes: Route[] = [
    { id: "28" },
    { id: "39" },
    { id: "71" },
    { id: "73" },
    { id: "111" },
  ]
  const selectedRouteIds: RouteId[] = ["28", "39"]

  const tree = renderer.create(
    <RoutePicker
      routes={routes}
      selectedRouteIds={selectedRouteIds}
      dispatch={emptyDispatch}
    />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test("renders a loading/empty state", () => {
  const tree = renderer.create(
    <RoutePicker
      routes={null}
      selectedRouteIds={[]}
      dispatch={emptyDispatch}
    />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test("clicking a route selects it", () => {
  const mockDispatch = jest.fn()
  const routes = [{ id: "id" }]

  const routePicker = shallow(
    <RoutePicker
      routes={routes}
      selectedRouteIds={[]}
      dispatch={mockDispatch}
    />
  )

  routePicker
    .find(".m-route-picker__route-list-button--deselected")
    .first()
    .simulate("click")

  expect(mockDispatch).toHaveBeenCalledWith(selectRoute("id"))
})

test("clicking a selected route deselects it", () => {
  const mockDispatch = jest.fn()
  const routes = [{ id: "id" }]

  const routePicker = shallow(
    <RoutePicker
      routes={routes}
      selectedRouteIds={["id"]}
      dispatch={mockDispatch}
    />
  )

  routePicker
    .find(".m-route-picker__route-list-button--selected")
    .first()
    .simulate("click")

  expect(mockDispatch).toHaveBeenCalledWith(deselectRoute("id"))
})

test("clicking in the list of selected routes deselects a route", () => {
  const mockDispatch = jest.fn()
  const routes = [{ id: "id" }]

  const routePicker = shallow(
    <RoutePicker
      routes={routes}
      selectedRouteIds={["id"]}
      dispatch={mockDispatch}
    />
  )

  routePicker
    .find(".m-route-picker__selected-routes-button")
    .first()
    .simulate("click")

  expect(mockDispatch).toHaveBeenCalledWith(deselectRoute("id"))
})
