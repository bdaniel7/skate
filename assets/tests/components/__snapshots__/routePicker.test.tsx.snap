// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of routes 1`] = `
<div
  className="m-route-picker"
>
  <h2>
    Selected Routes
  </h2>
  <ul>
    <li>
      <button
        className="m-route-picker__selected-routes-button"
        onClick={[Function]}
      >
        28
      </button>
    </li>
    <li>
      <button
        className="m-route-picker__selected-routes-button"
        onClick={[Function]}
      >
        39
      </button>
    </li>
  </ul>
  <h2>
    Routes
  </h2>
  <ul>
    <li>
      <button
        className="m-route-picker__route-list-button--selected"
        onClick={[Function]}
      >
        *
        28
      </button>
    </li>
    <li>
      <button
        className="m-route-picker__route-list-button--selected"
        onClick={[Function]}
      >
        *
        39
      </button>
    </li>
    <li>
      <button
        className="m-route-picker__route-list-button--deselected"
        onClick={[Function]}
      >
        71
      </button>
    </li>
    <li>
      <button
        className="m-route-picker__route-list-button--deselected"
        onClick={[Function]}
      >
        73
      </button>
    </li>
    <li>
      <button
        className="m-route-picker__route-list-button--deselected"
        onClick={[Function]}
      >
        111
      </button>
    </li>
  </ul>
</div>
`;

exports[`renders a loading/empty state 1`] = `
<div
  className="m-route-picker"
>
  <h2>
    Selected Routes
  </h2>
  none
  <h2>
    Routes
  </h2>
  loading...
</div>
`;
